docs/Middleware-Guide.md

```diff
@@ -0,0 +1,82 @@
+# Introduction
```

Comment:
we should link to this doc from our top-level readme and the gitbook

Reply:
Thats the plan, once its merged